internal/stacks/stackaddrs/removed.go

```diff
@@ -4,134 +4,389 @@
+				index, ok := nextTraversal[2].(hcl.TraverseIndex)
+				if !ok {
+					// This is an error, with exactly 3 we don't have another
+					// traversal to go to after this so the last entry must
+					// be the index.
+					return RemovedFrom{}, diags.Append(&hcl.Diagnostic{
+						Severity: hcl.DiagError,
+						Summary:  "Invalid 'from' attribute",
+						Detail:   "The 'from' attribute must designate a component that has been removed, in the form `component.component_name` or `component.component_name[\"key\"].",
+						Subject:  expr.Range().Ptr(),
+					})
+				}
+
+				currentTraversal = nextTraversal
+				nextTraversal = nil
+				indexExpr = hcl.StaticExpr(index.Key, index.SrcRange)
+
+			default: // len(nextTraversal) > 3
+				if index, ok := nextTraversal[2].(hcl.TraverseIndex); ok {
+					currentTraversal = nextTraversal[:3]
+					nextTraversal = nextTraversal[3:]
+					indexExpr = hcl.StaticExpr(index.Key, index.SrcRange)
+					break
+				}
+				currentTraversal = nextTraversal[:2]
+				nextTraversal = nextTraversal[2:]
+			}
+
+			var name string
+
+			switch root := currentTraversal[0].(type) {
+			case hcl.TraverseRoot:
+				name = root.Name
+			case hcl.TraverseAttr:
+				name = root.Name
+			default:
+				return RemovedFrom{}, diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Invalid 'from' attribute",
+					Detail:   "The 'from' attribute must designate a component that has been removed, in the form `component.component_name` or `component.component_name[\"key\"].",
+					Subject:  expr.Range().Ptr(),
+				})
+			}
+
+			switch name {
+			case "component":
+				name, ok := currentTraversal[1].(hcl.TraverseAttr)
+				if !ok {
+					return RemovedFrom{}, diags.Append(&hcl.Diagnostic{
+						Severity: hcl.DiagError,
+						Summary:  "Invalid 'from' attribute",
+						Detail:   "The 'from' attribute must designate a component that has been removed, in the form `component.component_name` or `component.component_name[\"key\"].",
+						Subject:  expr.Range().Ptr(),
+					})
+				}
+
+				if len(nextTraversal) > 0 || current.Rest != nil {
+					return RemovedFrom{}, diags.Append(&hcl.Diagnostic{
+						Severity: hcl.DiagError,
+						Summary:  "Invalid 'from' attribute",
+						Detail:   "The 'from' attribute must designate a component that has been removed, in the form `component.component_name` or `component.component_name[\"key\"].",
+						Subject:  expr.Range().Ptr(),
+					})
+				}
+
+				removedFrom.Component = ComponentRemovedFrom{
+					Name:  name.Name,
+					Index: indexExpr,
+				}
+				return removedFrom, diags
+			case "stack":
+				name, ok := currentTraversal[1].(hcl.TraverseAttr)
+				if !ok {
+					return RemovedFrom{}, diags.Append(&hcl.Diagnostic{
+						Severity: hcl.DiagError,
+						Summary:  "Invalid 'from' attribute",
+						Detail:   "The 'from' attribute must designate a component that has been removed, in the form `component.component_name` or `component.component_name[\"key\"].",
```

Comment:
the error seem due to an invalid stack address. Perhaps the error message should mention the stack